### ISSUE TITLE
Chapter 6: SELECTing w/ JOIN by state_fips or county_fips gives inaccurate results

### ISSUE THREAD

**Issue by CactusKurt** [anthonydb/practical-sql#6]
SELECT c2010.geo_name,
      /snip
      FROM us_counties_2010 c2010 INNER JOIN us_counties_2000 c2000
**ON c2010.state_fips = c2000.state_fips**
   **AND c2010.county_fips = c2000.county_fips**
   AND c2010.p0010001 <> c2000.p0010001
ORDER BY pct_change DESC;

Verify the problem by...

SELECT geo_name, state_us_abbreviation
FROM us_counties_2000
WHERE state_fips = '1';

produces 0 results, but

SELECT geo_name, state_us_abbreviation
FROM us_counties_2010
WHERE state_fips = '1';

produces 67 results

The problem arises with single-digit fips codes.  Need to add a leading "0" ('01') to produce correct results for us_counties_2000 database.  I copied/pasted the code to produce/query the databases.

**Comment by anthonydb** [comment]
Hi,

I can't replicate this issue. I did a fresh install of PostgreSQL 10.4 on Windows 10 and a fresh download of the `us_counties_2010.csv` data file from this repo.

The leading zeros in what becomes the `state_fips` column in `us_counties_2010` are present in all rows in the `us_counties_2010.csv` file in the repo. After import, running this query shows 67 results for `state_fips` code `01`:

```
SELECT state_fips, count(*)
FROM us_counties_2010
GROUP BY state_fips
ORDER BY state_fips;
```

If by any chance you opened the CSV in Excel or another app (aside from a plain text editor) before importing it, it's possible the leading zeros were stripped. This is common (and I cover the scenario in the book).

Please drop the table, download a fresh copy of the file, and try the import again.

**Comment by CactusKurt** [comment]
I opened both files in a spreadsheet app to view them before COPYing them.  If you look at my original comment, the problem was how the query was handled differently by the us_counties_2000 and the us_counties_2010 databases.  Using a single digit "1" in us_counties_2000 produced 0 results, but using a single digit "1" in us_counties 2010 produced 67 results.  Contrarily, using a 2-digit "01" in us_counties_2000 produced 67 results, but a 2-digit "01" in us_counties_2010 produced 0 results.

It didn't make any sense.  It also caused the JOIN ON in your example to fail to include states with single digit fips codes.  Apparently, the leading "0" was stripped from the state_fips data in the us_counties_2010 as you suggested (but why wasn't it stripped in the us_counties_2000 file?)

This is very confusing and non-intuitive for a beginner (like me).  Opening the file in a spreadsheet for viewing will alter the data?

After dropping both tables and downloading fresh copies (without opening in a spreadsheet), everything seems to work as expected.  A state_fips code of "01" produces the correct results (67) for both files.

**Comment by anthonydb** [comment]
Hi @CactusKurt,

Unfortunately, Excel will indeed strip leading zeros from numbers in a CSV file if you don't explicitly tell it not do while doing an import. See https://support.office.com/en-us/article/Keeping-leading-zeros-and-large-numbers-1bf7b935-36e1-4985-842f-5dfa51f85fe7

I discuss this in Practical SQL on page 135, in the section "Checking for Malformed Values Using length()". You're not the first person to experience this -- I messed up data this way when I started out, and I daresay this behavior of Excel's has bitten more than a few people.

Your note prompts me to think I should include a warning earlier in the book and also here on GitHub, which I will do.

Did importing the files without opening them first in Excel also solve https://github.com/anthonydb/practical-sql/issues/5 ?

**Comment by CactusKurt** [comment]
No, the problem in #5 sill persisted after dropping the table us_counties_2010 and using a clean copy (not opened in Excel).  Still getting the same error. I deleted the tilde character in a text editor and it resolved the issue.
